src/lang/typing.ml

```diff
@@ -165,97 +169,183 @@ let rec occur_check a b =
-                        if not (List.mem Dtools c) then
-                          b.descr <- EVar (j, Dtools :: c)
-                    | _ -> raise (Unsatisfied_constraint (Dtools, b)))
-              | InternalMedia -> (
-                  let is_internal name =
-                    try
-                      let kind = Frame_content.kind_of_string name in
-                      Frame_content.is_internal kind
-                    with Frame_content.Invalid -> false
-                  in
-                  match b.descr with
-                    | Constr { name } when is_internal name -> ()
-                    | Ground (Format f)
-                      when Frame_content.(is_internal (kind f)) ->
-                        ()
-                    | EVar (j, c) ->
-                        if List.mem InternalMedia c then ()
-                        else b.descr <- EVar (j, InternalMedia :: c)
-                    | _ -> raise (Unsatisfied_constraint (InternalMedia, b)))
-              | Num -> (
-                  match (demeth b).descr with
-                    | Ground g ->
-                        if g <> Int && g <> Float then
-                          raise (Unsatisfied_constraint (Num, b))
-                    | EVar (j, c) ->
-                        if List.mem Num c then ()
-                        else b.descr <- EVar (j, Num :: c)
-                    | _ -> raise (Unsatisfied_constraint (Num, b))))
+                          if not (List.mem Dtools c) then
+                            b.descr <- EVar (j, Dtools :: c)
+                      | _ -> raise (Unsatisfied_constraint (Dtools, b)))
+                | InternalMedia -> (
+                    let is_internal name =
+                      try
+                        let kind = Frame_content.kind_of_string name in
+                        Frame_content.is_internal kind
+                      with Frame_content.Invalid -> false
+                    in
+                    match b.descr with
+                      | Constr { name } when is_internal name -> ()
+                      | Ground (Format f)
+                        when Frame_content.(is_internal (kind f)) ->
+                          ()
+                      | EVar (j, c) ->
+                          if List.mem InternalMedia c then ()
+                          else b.descr <- EVar (j, InternalMedia :: c)
+                      | _ -> raise (Unsatisfied_constraint (InternalMedia, b)))
+                | Num -> (
+                    match (demeth b).descr with
+                      | Ground g ->
+                          if g <> Int && g <> Float then
+                            raise (Unsatisfied_constraint (Num, b))
+                      | EVar (j, c) ->
+                          if List.mem Num c then ()
+                          else b.descr <- EVar (j, Num :: c)
+                      | _ -> raise (Unsatisfied_constraint (Num, b))))
+              constraints;
             constraints
```

Comment:
This a bit weird, you're doing: 
```ocaml
let constraints = 
  List.iter fn constraints;
  constraints
in
```
I feel like this could be rewritten.

Reply:
excepting that you need the match to extract the `constraints`!